### DISCUSSION TITLE
Lab 06.02 Task 3

### DISCUSSION THREAD

**Discussion by moralesluis26** [rramosp/ai4eng.v1#198] [06 - MÉTODOS NO SUPERVISADOS]
Saludos.
Tengo una duda con este ejercicio, y es que ya realicé el código tal como lo dice la descripción y utilicé el código anterior, pero según la calificación tengo algo mal que aún no encuentro.
De antemano le agradezco la atención.

def find_similar(v,X,n):
  from sklearn.decomposition import NMF 
  from numpy import linalg as la
  nmf = NMF(n_components=30, init="random", random_state=0)
  nmf = nmf.fit(X) ## your code here. call the 'fit' method
  Xt = nmf.fit_transform(X) # use nmf to transform X
  vt = nmf.fit_transform(v.reshape(1, -1)) # use nmf to transform v .. you will have to use reshape like this v.reshape(1,-1)
  assert n<len(X), "n must at most the number of vectors in X"
  from numpy import linalg as la
  r = np.array([la.norm(vt-Xt[i]) for i in range(len(Xt))])
  t = np.argsort(r)
  result= t[0:n]
  return result

Imagén de los resultados:
![imagen](https://user-images.githubusercontent.com/46792756/188281718-0c3abd28-3231-42ca-a4b9-3567645ae0c7.png)

**Comment by Felipe-RA** [comment]
Saludos MoralesLuis

Disculpa la tardanza, ¡abrí tu post hace tres días y pensé que ya lo había respondido!

Primero, recordarte utilizar la estructura de bloques de código para hacer más legible tus preguntas:

```
def find_similar(v,X,n):
    from sklearn.decomposition import NMF
    from numpy import linalg as la
    nmf = NMF(n_components=30, init="random", random_state=0)
    nmf = nmf.fit(X) ## your code here. call the 'fit' method
    Xt = nmf.fit_transform(X) # use nmf to transform X
    vt = nmf.fit_transform(v.reshape(1, -1)) # use nmf to transform v .. you will have to use reshape like this v.reshape(1,-1)
    assert n<len(X), "n must at most the number of vectors in X"
    from numpy import linalg as la
    r = np.array([la.norm(vt-Xt[i]) for i in range(len(Xt))])
    t = np.argsort(r)
    result= t[0:n]
    return result
```

El problema está  en tu aplicación  de `closest`, que no retorna un arraylike válido con los indices requeridos para esta implementación.

Te recomiendo también para hacer más legible tu código (y que se vea mejor el flujo de las transformaciones) incluir en el código la función de `closest` en lugar de volcar su contenido en el código propio de `find_similarity`. Esto tiene el beneficio de que te permite identificar con mayor facilidad posibles problemas.

También, me gustaría recomendarte esta implementación de `closest`, la cual es más escueta y clara que la que intuyo de tu fragmento de código:
```
import numpy as np
from numpy import linalg as LA
result = np.argsort(LA.norm(v-X, axis=1))[:n]
```

Fijate como aquí te puedes ahorrar el `list comprehension`, la creación de un arreglo intermedio y el `slicing` del arreglo.

Cualquier duda me comentas!
